index.js

```diff
@@ -1,227 +1,261 @@
+/**
+ * Lock and cache using redis, memory, and more!
+ * @example
+ * const cache = require('lock_and_cache')
+ * let square = x => cache(x, x => x*x)
+ * @example
+ * const cache = require('lock_and_cache')
+ * let square = cache.wrap(x=>x*x)
+ * @example
+ * const lock_and_cache = require('lock_and_cache')
+ * let cache = lock_and_cache.LockAndCache()
+ * let square = cache.wrap(x=>x*x)
+ * let foo = cache.get('foo', ()=>'bar')
+ */
+
 var redis = require('redis')
 var Redlock = require('redlock')
 var assert = require('assert')
+var cacheManager = require('cache-manager')
+var redisStore = require('cache-manager-redis-store');
 
-const DEFAULT_REDIS_OPTIONS = {
-  retry_strategy (options) {
-    return Math.min(options.attempt * 100, 3000)
-  }
-}
+const LOCK_TIMEOUT = 5000
+const LOCK_EXTEND_TIMEOUT = 2500
 
 const ERROR_TTL = 1
 
-function defaults (opts, defs) {
-  let options = {}
-  for (let key in defs) {
-    options[key] = defs[key]
-  }
-  for (let key in opts) {
-    options[key] = opts[key]
-  }
-  return options
+const DEFAULT_REDIS_LOCK_OPTS = {
+  retry_strategy (options) {
+    return Math.min(options.attempt * 100, 3000)
+  },
+  url: (process.env.LOCK_URL || process.env.REDIS_URL),
 }
 
-function redisGet (client, key) {
-  return new Promise(function lookup (resolve, reject) {
-    client.get(key, function onGet (err, value) {
-      if (err) {
-        reject(err)
-      } else if (value === null) {
-        reject('key not found')
-      } else if (value === 'undefined') {
-        resolve(undefined)
-      } else {
-        resolve(JSON.parse(value))
-      }
-    })
-  })
+const DEFAULT_MEM_CACHE_OPTS = {
+      store: 'memory',
+      max: 10,
+      ttl: 30
 }
 
-function redisSet (client, key, ttl, value) {
-  return new Promise(function set (resolve, reject) {
-    var storedValue
-    if (typeof value === 'undefined') {
-      storedValue = 'undefined'
-    } else {
-      storedValue = JSON.stringify(value)
-    }
-    client.setex(key, ttl, storedValue, function onSet (err) {
-      if (err) {
-        reject(err)
-      } else {
-        resolve(value)
-      }
-    })
-  })
+const DEFAULT_REDIS_CACHE_OPTS = {
+      store: redisStore,
+      url: (process.env.CACHE_URL || process.env.REDIS_URL),
+      ttl: 600
 }
 
-function lockAndCache (getLocker, key, ttl, work) {
-  key = JSON.stringify(key)
 
-  var locker = getLocker()
-  var cacheClient = locker[0]
-  var redlock = locker[1]
-  var unref = locker[2]
+function defaults (opts, defs) {
+  return Object.assign({}, opts, defs);
+}
 
-  var lock
-  var value
-  var done
-  var extendTimeout
 
-  function extend () {
-    if (lock && !done) {
-      lock.extend(5000, function onExtend (err) {
-        if (lock && !done) {
-          extendTimeout = setTimeout(extend, err ? 500 : 2500)
-        }
-      })
-    }
+/**
+ * runs cb() then calls obj.close()
+ *
+ * @example
+ * closing(LockAndCache(), cache => console.log(cache.get('key', (val)=>{...})))
```

Comment:
this closes the connections while the function is still running?

I think requiring the user to wrap all use of the library in `closing`, or to manually run `close` is too big of an API change. Internally we depend quite a lot on the ability to use lock_and_cache without keeping track of connections.

Reply:
no, it closes after the callback returns

manual closing is optional; refcounting will close the connection eventually